lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -0,0 +1,215 @@
+                "arch": "noarch",
+                "epoch": null,
+                "name": "yum",
+                "release": "513.fc27",
+                "source": "rpm",
+                "version": "3.4.3"
+            }
+          ],
+          "yum-utils": [
+            {
+                "arch": "noarch",
+                "epoch": null,
+                "name": "yum-utils",
+                "release": "513.fc27",
+                "source": "rpm",
+                "version": "1.1.31"
+            }
+          ],
+        }
+    deb_packages:
+      description: list of dicts with package information
+      returned: when dpkg, apt, or aptitude is the specified or auto detected manager
+      type: dict
+      sample:
+        {
+          "libbz2-1.0": [
+          {
+              "version": "1.0.6-5",
+              "source": "apt",
+              "arch": "amd64",
+              "name": "libbz2-1.0"
+          }
+          ],
+          "patch": [
+          {
+              "version": "2.7.1-4ubuntu1",
+              "source": "apt",
+              "arch": "amd64",
+              "name": "patch"
+          }
+          ],
+        }
+'''
+
+import sys
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+
+# Alias mappings of different ways package namespaces might be referred to
+RPM_ALIASES = ('rpm', 'dnf', 'yum')
+DEB_ALIASES = ('apt', 'aptitude', 'dpkg')
+
+
+def rpm_package_list():
+
+    try:
+        import rpm
+    except ImportError:
+        module.fail_json('Unable to use the rpm python bindings, please ensure they are installed under the python the module runs under')
+
+    trans_set = rpm.TransactionSet()
+    installed_packages = {}
+    for package in trans_set.dbMatch():
+        package_details = dict(name=package[rpm.RPMTAG_NAME],
+                               version=package[rpm.RPMTAG_VERSION],
+                               release=package[rpm.RPMTAG_RELEASE],
+                               epoch=package[rpm.RPMTAG_EPOCH],
+                               arch=package[rpm.RPMTAG_ARCH],
+                               source='rpm')
+        if package_details['name'] not in installed_packages:
+            installed_packages[package_details['name']] = [package_details]
+        else:
+            installed_packages[package_details['name']].append(package_details)
+    return installed_packages
+
+
+def deb_package_list():
+
+    try:
+        import apt
+    except ImportError:
+        module.fail_json('Unable to use the apt python bindings, please ensure they are installed under the python the module runs under')
+
+    apt_cache = apt.Cache()
+    installed_packages = {}
+    apt_installed_packages = [pk for pk in apt_cache.keys() if apt_cache[pk].is_installed]
+    for package in apt_installed_packages:
+        ac_pkg = apt_cache[package].installed
+        package_details = dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, source='apt')
```

Comment:
It would be easier if RPM and DEB were to give the same structured reply, or having an agnostic module would not be very useful (in cases like "if upstream version is greater than x.y"). Comparing versions is a complex operation and there is no filter around `dpkg --compare-versions`.

Also having the full version (epoch+upstream+release, following the standard distro notation) along with the upstream version would be useful I think.

Reply:
So this was my comment from #32871 and as this part of the code did not change it seemed still appropriate. With package namespacing, which can be nice if we want to later support language-specific package manager (pip…), as long as the proposed structure makes it easy to use `ansible_pkg_mgr` to find the "main" package manager information, then this is still useful (version comparison and so on).

Reply:
My main concern here with trying to comply with a standard reply structure is when this is eventually extended to all sorts of package managers such as pip, rubygems, maven, ant, npm, cpan, pecl, pear, homebrew, etc etc. What happens when one of those providers doesn't offer the fields we set in the standard response? And what do we do with the extra data outside the bounds of the standard?

Reply:
I would settle on the 'lowest common denominator' as the standard (document in return fields) and make notes about additional fields available depending on package  manager/distro

Reply:
Lowest common denominator seems to be `name-version` if we include the language package managers but that's effectively useless when comparing distro packages so I'm not sure how useful it would be in the end.

We could have two classifications: `name-version-release` for all distro level packages and then `name-version` for all others.Thoughts?

Reply:
It seems fine.

In this case I think the next step would be to provide a filter able to split this information according to the package manager name and `name-version-release`. This filter would only make sense if used with distro-level results because they have the same concepts of upstream version/distro release number/…. It could be reused outside the scope of this PR, which would be nice.

This could be done in a separate PR though.

Reply:
@duck-rh I don't follow what you mean by a filter, why would we be filtering these here? Seems out of scope for a fact gathering module.

Reply:
@bcoca thoughts on the `name-version-release` format for distro packages and `name-version` for "all others"?

Reply:
Instead of include the 'source' in the package info dict, could the return data have the 'source' as the first level?

That way deb/rpm/gem/mypkgformat data schema dont have to match each other exactly, as long as the items inside those keys are the same. (The more consistent and similar they are the better of course, but...)

ie, instead of:

``` json
{
          "yum": [
            {
                "arch": "noarch",
                "epoch": null,
                "name": "yum",
                "release": "513.fc27",
                "source": "rpm",
                "version": "3.4.3"
            }
          ],
          "yum-utils": [
            {
                "arch": "noarch",
                "epoch": null,
                "name": "yum-utils",
                "release": "513.fc27",
                "source": "rpm",
                "version": "1.1.31"
            }
          ],
        }
```

something like:

``` json
{
    "rpm": {
              "yum": [
                {
                    "arch": "noarch",
                    "epoch": null,
                    "name": "yum",
                    "release": "513.fc27",
                    "version": "3.4.3"
                }
              ],
              "yum-utils": [
                {
                    "arch": "noarch",
                    "epoch": null,
                    "name": "yum-utils",
                    "release": "513.fc27",
                    "version": "1.1.31"
                }
              ],
            },
    "deb": {
             "somedeb": [
                 {
                     "name": "somedeb",
                     "version": "1.1.1-1",
                     "arch": "amd64"
                 }
              ],
             "otherdeb": [
                 {
                     "name": "otherdeb",
                     "version": "2.2.2-2",
                     "arch": "amd64"
                 }
             ]
    }
}
```

Reply:
We are effectively getting this structure as it stands now, we'll return the `ansible_facts.packages.rpm` or `ansible_facts.packages.deb` and the user will be presented with this namespaced structure as you describe within `ansible_facts.packages` ... if it is preferred to strip out the `source` from the `rpm` entries, that's fine but I think the end result will be the same since you can only gather information from a single package manager per task execution and they are aggregated under the `ansible_facts.packages` dict.

Reply:
@maxamillion Excellent. I missed that there was already a per 'manager' sub dict in the return, so it already does what I was suggesting so that comment can be ignored ;->